File: PythonVersion/QuickUnion.py
#
#Quick Union features
#
def find_root_QU(initial_id, i):#find the root of node i
    while (i != initial_id[i]):
        i = initial_id[i]
    return i #return the root

def connected_QU(initial_id,p,q):#find if two nodes have the same root
    return(find_root_QU(initial_id,p)==find_root_QU(initial_id,q))

def union_QU(initial_id, p, q):#list,node1,node2
    local_id_list = initial_id
    local_i = find_root_QU(initial_id,p)
    local_j = find_root_QU(initial_id,q)
    local_id_list[local_i] = local_j
    return local_id_list

File: C++Version/quickunion.cpp
#include <vector>

using namespace std;

int findroot_QU(vector<int> vec, int x){
	int ans = x;
	while (ans != vec[ans])
		ans = vec[ans];
	return ans;
}

int connected_QU(vector<int> vec, int x, int y){
	return (findroot_QU(vec, x) == findroot_QU(vec, y));
}

void union_QU(vector<int> &vec, int x, int y){
	int root_x = findroot_QU(vec, x);
	int root_y = findroot_QU(vec, y);
	vec[root_x] = root_y;
}
